### PR TITLE
Move RegionKind to rustc_type_ir

### PR DESCRIPTION
**Pull request by jackh726**
(Also UniverseIndex)

r? rust-lang/types

### CODE REVIEW

**Comment by compiler-errors** [comment]
Looks great! I'll give it another look-over once perf gets back and CI looks green, but glad to see that this was a much simpler move than migrating TyKind.

**Comment by jackh726** [comment]
> Looks great! I'll give it another look-over once perf gets back and CI looks green, but glad to see that this was a much simpler move than migrating TyKind.

So much simpler! The worst part was writing the impls 🙃

**Comment by compiler-errors** [comment]
Oh, did you write them by hand? When I made them for TyKind, I just got them from cargo-expand + a lot of string replacement to clean them up. If you did write them by hand, I'll try to give them a particularly close looking over for correctness.

**Comment by jackh726** [comment]
> Oh, did you write them by hand? When I made them for TyKind, I just got them from cargo-expand + a lot of string replacement to clean them up. If you did write them by hand, I'll make sure to give them a particularly close looking over for correctness.

Yeah, I did. Probably would have been roughly the same time to do the other way.

File: compiler/rustc_type_ir/src/sty.rs [hunk 1]
```diff
@@ -991,3 +991,439 @@ where
...
+        ReStatic => 3,
+        ReVar(_) => 4,
+        RePlaceholder(_) => 5,
+        ReEmpty(_) => 6,
+        ReErased => 7,
+    }
+}
+
+// This is manually implemented because a derive would require `I: Copy`
+impl<I: Interner> Copy for RegionKind<I>
```
Comment (eggyal) [inline compiler/rustc_type_ir/src/sty.rs:1155]: Does it matter that `derive` would introduce that bound?  In chalk, `Copy` is a supertrait of `Interner`; and in rustc, isn't `I` just a shared ref to the type context (which therefore is indeed `Copy`)?

I guess the same should apply for `Clone` even though it is not strictly a supertrait of `Interner`?  After all, requiring `I: Clone` shouldn't be restrictive if we know that `I: Copy`.

As an aside, I wonder whether there wouldn't be some ergonomic benefit to using a `derivative`-like macro here to workaround #26925 without requiring all these manual impls?

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
@bors r+ 

:tada:

**Pull request merged**